include/envoy/network/io_handle.h

```diff
@@ -7,27 +7,93 @@
 namespace Envoy {
 namespace Network {
 
+namespace Address {
+class Instance;
+} // namespace Address
+
+/**
+ * Basic type for return result which has a return code and error code defined
+ * according to different implementation.
+ */
+template <typename T> struct IoHandleCallResult {
+  virtual ~IoHandleCallResult() {}
+
+  T rc_;
+  int errno_;
+};
+
+typedef IoHandleCallResult<int> IoHandleCallIntResult;
+typedef IoHandleCallResult<ssize_t> IoHandleCallSizeResult;
+
 /**
  * IoHandle: an abstract interface for all I/O operations
  */
 class IoHandle {
 public:
-  IoHandle() {}
-
   virtual ~IoHandle() {}
 
   /**
    * Return data associated with IoHandle.
    *
    * TODO(sbelair2) remove fd() method
+   * We probably still need some method similar to this one for
+   * evconnlistener_new(). Or We can move it to IoSocketHandle and down cast the
+   * IoHandle to IoSocketHandle wherever needed.
    */
   virtual int fd() const PURE;
 
   /**
    * Clean up IoHandle resources
    */
-  virtual void close() PURE;
+  virtual IoHandleCallIntResult close() PURE;
+
+  virtual bool isClosed() PURE;
+
+  virtual IoHandleCallSizeResult readv(const iovec* iovec, int num_iovec) PURE;
+
+  virtual IoHandleCallSizeResult writev(const iovec* iovec, int num_iovec) PURE;
+
+  virtual IoHandleCallIntResult bind(const Network::Address::Instance& address) PURE;
+
+  virtual IoHandleCallIntResult connect(const Network::Address::Instance& server_address) PURE;
+
+  /**
+   * Wrap setsockopt()
+   */
+  virtual IoHandleCallIntResult setIoHandleOption(int level, int optname, const void* optval,
```

Comment:
I would suggest we use an enum for the options we care about. The one gotcha here is IIRC we have some capability in the config to actually set options that the code doesn't know about, by int, so maybe this isn't possible and we need to keep this? Thoughts?

Reply:
Would there be potential crash if envoy fails to map the value set in config and the enum defined here? If so, introducing enum for socket option can be error-prone.

Reply:
Right I think if we have a config option to set an int, we need to leave it as is.